Local recipes index post (#266)
conan-io/conan-io.github.io: _posts/2024-04-23-Introducing-local-recipes-index-remote.markdown

---
layout: post
comments: false
title: "Introducing the Local-Recipes-Index: A New Repository Type in Conan"
meta_title: "Understanding Conan's Local-Recipes-Index to improve package management for C/C++ - Conan Blog"
description: "Explore the capabilities of Conan's local-recipes-index feature and how it enhances dependency management for C and C++ libraries."
---

ConanCenter, the central open-source repository for C and C++ libraries, is a success

> story, containing a vast majority of C and C++ open-source packages. It processes hundreds

Reviewer: either `the vast majority` or `a vast collection`

> story, containing a vast majority of C and C++ open-source packages. It processes hundreds

Author: ```suggestion
story, containing a vast collection of C and C++ open-source packages. It processes hundreds
```

of thousands of download requests daily, building its packages [from a central GitHub
repository](https://github.com/conan-io/conan-center-index), which is organized in a

> specific folder layout and has received nearly 6,000 pull requests from contributors in

Reviewer: ```suggestion
specific folder layout and received nearly 6,000 pull requests from contributors in
```

binaries downloaded from the internet. Instead, they build their own binaries in-house
using the `conan-center-index` recipes. These organizations often need to customize these
recipes to meet unique requirements that are not applicable to the broader community,
making such contributions unsuitable for the upstream repository. The Conan documentation
[acknowledges this
practice](https://docs.conan.io/2/devops/conancenter/hosting_binaries.html), recommending
working from a fork of `conan-center-index` for such needs.

For all those reasons in version 2.2.0, Conan introduced a new repository type called
`local-recipes-index`, designed to offer more flexibility by allowing a Conan remote to
point to a local copy of Conan recipes with a specific layout.

The `local-recipes-index` allows users to maintain a local folder with the same structure
as the `conan-center-index` GitHub repository, using it as a source for package recipes.
This new type of repository is recipes-only, necessitating the construction of package
binaries from source on each machine where the package is used. For sharing binaries
across teams, we continue to recommend [using a Conan remote server like

> Artifactory](https://www.google.com/url?q=https://docs.conan.io/2/tutorial/conan_repositories/setting_up_conan_remotes/artifactory/artifactory_ce_cpp.html%23artifactory-ce-cpp&sa=D&source=docs&ust=1713353795928390&usg=AOvVaw1tq4PVvg6umcEqZUqMY4Lo)

Reviewer: Google url?

> Artifactory](https://www.google.com/url?q=https://docs.conan.io/2/tutorial/conan_repositories/setting_up_conan_remotes/artifactory/artifactory_ce_cpp.html%23artifactory-ce-cpp&sa=D&source=docs&ust=1713353795928390&usg=AOvVaw1tq4PVvg6umcEqZUqMY4Lo)
> for production purposes.

Author: ```suggestion
Artifactory](https://docs.conan.io/2/tutorial/conan_repositories/setting_up_conan_remotes/artifactory/artifactory_ce_cpp.html)
for production purposes.
```

- This feature enables contributors to share package recipes with the community for
  libraries that might not be suitable for ConanCenter due to various reasons, such as
  licensing constraints or binary distribution policies. An example of this could be CUDA
  or other proprietary libraries, which are distributed as precompiled closed-source
  binaries.

- It simplifies the adoption of best practices outlined [in the Conan
  documentation](https://docs.conan.io/2/devops/conancenter/hosting_binaries.html) for
  organizations requiring custom-built binaries or modified ConanCenter recipes to meet
  unique requirements. This approach grants users complete control over their third-party
  dependencies, ensuring they are both robust and fully customizable.

Next, we'll delve into practical examples to demonstrate these two use cases of the
`local-recipes-index` repository.

## Using a local-recipes-index repository with your own recipes

In this section, we will illustrate how to use the `local-recipes-index` feature for
scenarios where certain libraries or tools, due to licensing restrictions or proprietary
nature, are not suitable for ConanCenter.

> For demonstration, let's create a `local-recipes-index` repository for a hypothetical

Reviewer: ```suggestion
For demonstration purposes, let's create a `local-recipes-index` repository for a hypothetical
```

Now you can list and install packages from this new repository:

    $ conan list "*" -r=mylocalrepo
    $ conan install --requires=hello/0.1 -r=mylocalrepo --build=missing

At this point, you could push this repository to your GitHub account and share it with the
community. Please be aware that, as we commented earlier, this feature is specifically
tailored for scenarios where certain libraries are not suitable for ConanCenter. Remember,
a "local-recipes-index" repository has limitations: it is not fully reproducible as it
models only versions and not revisions, and it does not provide binaries. Therefore,
outside of these cases, it is advised to use a remote package server such as Artifactory.

Now, users simply need to clone the GitHub repository and add the cloned folder as a local
repository themselves.

## Building Binaries from a private `conan-center-index` fork

As outlined in [the Conan DevOps
Guide](https://docs.conan.io/2/devops/using_conancenter.html), there are many cases where
organizations need to operate independently of ConanCenter by building their own binaries.
Being decoupled from the public upstream ConanCenter server and building your own binaries

> from a fork of ``conan-center-index`` as suggested in that documentation page can have

Reviewer: ```suggestion
from a fork of ``conan-center-index`` as suggested in the linked documentation page can have
```

many advantages, including absolute control and possibility to customize recipes, giving
us the ability for the repository to act as a snapshot of versions, be completely robust
against possible continuous changes and new releases in upstream ConanCenter, etc.

The `local-recipes-index` repository allows you to easily build binaries from a fork of
`conan-center-index`, and then hosting them on a Conan remote repository like Artifactory.
The main difference with the process explained [in the Conan DevOps
guide](https://docs.conan.io/2/devops/conancenter/hosting_binaries.html) is the ability
to immediately test multiple local changes without the need to export each time a recipe
is modified.

> Note that in this case, mixing binaries from ConanCenter with locally built binaries is
> not recommended mixing binaries from ConanCenter with locally built binaries is not
> recommended for several reasons:

Reviewer 2: ```suggestion
Note that in this case, mixing binaries from ConanCenter with locally built binaries is
not recommended for several reasons:
```

- Binary compatibility: There may be small differences in setup between the ConanCenter CI

> and user CI. Maintaining a consistent setup for all binaries can mitigate some issues.

Reviewer: ```suggestion
and the user's CI. Maintaining a consistent setup for all binaries can mitigate some issues.
```

- Full control over builds: Building all binaries yourself ensures you have complete control
over the compilation environment and dependency versions.

> Instead, build all your direct and transitive dependencies from the fork.

Reviewer: ```suggestion
Instead, it's recommended to build all your direct and transitive dependencies from the fork.
```

> To begin, remove the upstream ConanCenter as it will not be used, everything will come

Reviewer: ```suggestion
To begin with, remove the upstream ConanCenter remote as it will not be used, everything will come
```

our organization, users and CI jobs:

    $ conan remote add myartifactoryrepo <artifactory_url>
    $ conan upload zlib* -r=myartifactoryrepo -c

This way, consumers of the packages will not only enjoy the pre-compiled binaries and
avoid having to always re-build from source all dependencies, but that will also provide
stronger guarantees that the dependencies build and work correctly, that all dependencies
and transitive dependencies play well together, etc. Decoupling the binary creation
process from the binary consumption process is the way to achieve faster and more reliable
usage of dependencies.

Remember, in a production setting, the `conan upload` command should be executed by CI,
not developers, following the [Conan
guidelines](https://docs.conan.io/2/knowledge/guidelines.html). This approach ensures that
package consumers enjoy pre-compiled binaries and consistency across dependencies.

### Modifying the local-recipes-index repository files

One of the advantages of this approach is that all the changes that we do in every single
recipe are automatically available for the Conan client. For example, changes to the

> `recipes/zlib/config.yml` file, are immediately recognized by the Conan client. If you

Reviewer: ```suggestion
`recipes/zlib/config.yml` file are immediately recognized by the Conan client. If you
```

> edit that file and remove all versions but the latest and then we `list` the recipes:

Reviewer: ```suggestion
edit that file and remove all versions but the latest and then call `conan list`, the returned recipes are:
```

  replacement for Conan's remote package servers, which play a vital role in hosting
  packages for regular use.

- Also, note that a server remote can retain a history of changes storing multiple recipe
  revisions. In contrast, a `local-recipes-index` remote can only represent a single
  snapshot at any given time. 

Furthermore, this feature does not support placing server URLs directly in recipes; remote
repositories must be explicitly added with `conan remote add`. Decoupling abstract package
requirements, such as "zlib/1.3.1", from their specific origins is crucial to resolving
dependencies correctly and leveraging Conan's graph capabilities, including version
conflict detection and resolution, version-ranges resolution, [opting into
pre-releases](https://docs.conan.io/2/devops/versioning/resolve_prereleases.html),
[platform_requires](https://docs.conan.io/2/reference/config_files/profiles.html#platform-requires),
[replace_requires](https://docs.conan.io/2/reference/config_files/profiles.html#replace-requires),
etc. This separation also facilitates the implementation of modern DevOps practices, such

> as package immutability and full relocatability and package promotions.

Reviewer 2: ```suggestion
as package immutability, full relocatability and package promotions.
```
